### PR TITLE
Added 'Floating' field to Node

### PR DESCRIPTION
**Pull request by ivanmilov**
[i3 PR #4387](https://github.com/i3/i3/pull/4387)

### CODE REVIEW

**Comment by ivanmilov** [comment]
Is [TestRestartSubprocess](https://github.com/i3/go-i3/blob/master/restart_test.go#L15-L40) broken?
It seems the ['t'](https://github.com/i3/go-i3/blob/master/restart_test.go#L39)  should not be present in a goroutine
```go
func TestRestartSubprocess(t *testing.T) {
  go func() {
    ...
    t.Fatal(recv.Close()) // should not be reached
  }
  ...
}
```

**Comment by ivanmilov** [comment]
recheck

File: tree.go [hunk 1]
```diff
@@ -74,6 +74,17 @@ const (
 	FullscreenGlobal FullscreenMode = 2
 )
 
+// FloatingType indicates the floating type of Node.
+type FloatingType string
+
+// i3 currently implements the following node types:
+const (
+	AutoOff FloatingType = "auto_aff"
```
Comment (stapelberg) [inline tree.go:82]: typo: auto_off

File: tree.go [hunk 2]
```diff
@@ -99,6 +110,7 @@ type Node struct {
 	Focus              []NodeID         `json:"focus"`
 	Nodes              []*Node          `json:"nodes"`
 	FloatingNodes      []*Node          `json:"floating_nodes"`
+	Floating           string           `json:"floating"`
```
Comment (stapelberg) [inline tree.go:113]: Shouldn’t this field be of type FloatingType?

File: tree.go (hunk 1 above)
Comment (ivanmilov) [reply tree.go]: fixed

File: tree.go (hunk 2 above)
Comment (ivanmilov) [reply tree.go]: fixed

**Comment by ivanmilov** [comment]
@stapelberg As you are here, let me ask a question :)
I have a couple of helper functions for the tree and this PR is a prerequisite for it. Are you open to having some extra functionality here?
Currently, I have these functions as helpers:
```golang
func (n *Node) FindFocusedLeaf() *Node 
func (child *Node) FindParent() *Node 
func (n *Node) IsFloating() bool 
```

**Comment by stapelberg** [comment]
> Currently, I have these functions as helpers:

I’m not sure if these are pulling their weight? There’s also always the concern of introducing too much API surface. Maybe helpers can go into a separate package (in the same repository)?

In one of my programs, I’m using this code, which I think is FindFocusedLeaf().FindParent() in your helpers?

```go
	// Get the parent of the currently focused container.
	parent := tree.Root.FindFocused(func(n *i3.Node) bool {
		for _, c := range n.Nodes {
			if c.Focused {
				return true
			}
		}
		return false
	})
```

I would be curious to see if there are helpers that virtually everyone needs, or at least most programs.

My own programs are too different from each other thus far to make such an observation…

**Review by stapelberg — APPROVED** [review]

**Comment by ivanmilov** [comment]
>I would be curious to see if there are helpers that virtually everyone needs, or at least most programs.

I got your point. Currently, I have 3 projects and all of them need these functions. Basically, all my helpers are:

```golang
package i3

import "strings"

func (n *Node) FindFocusedLeaf() *Node {
	return n.FindFocused(func(n *Node) bool { return n.Focused })
}

func (child *Node) FindParent() *Node {
	tree, err := GetTree()
	if err != nil {
		return nil
	}
	parent := tree.Root.FindChild(func(n *Node) bool {
		for _, f := range n.Focus {
			if f == child.ID {
				return true
			}
		}
		return false
	})

	return parent
}

func (n *Node) IsFloating() bool {
	if len(n.Floating) < 4 {
		return false
	}
	return strings.HasSuffix(string(n.Floating), "_on")
}

```

>Maybe helpers can go into a separate package (in the same repository)?

In this case, as far as I know,  it is impossible to use 'receiver methods' like _func **(t Type)** FunctionName()..._

**Comment by stapelberg** [comment]
> In this case, as far as I know, it is impossible to use 'receiver methods' like _func **(t Type)** FunctionName()..._

Yes, but you can just pass the Node as a parameter, or define your own type:
```
type MyNode i3.Node
func (n *MyNode) …

// MyNode(tree.Root).FindFocusedLeaf()
```

**Pull request merged**